src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1306,7 +1305,23 @@ public void checkout(Run<?, ?> build, Launcher launcher, FilePath workspace, Tas
             ext.beforeCheckout(this, build, git, listener);
         }
 
-        retrieveChanges(build, git, listener);
+        for (int tryCount = 1; tryCount <= 5; tryCount++) {
+            try {
+                retrieveChanges(build, git, listener);
+                break;
+            } catch (GitException ex) {
+                if (tryCount == 5) {
+                    ex.printStackTrace(listener.error(ex.getMessage()));
+                    throw new AbortException(ex.getMessage());
+                } else {
+                    int waitTime = tryCount * 10;
+                    listener.getLogger().println("Git retrieve changes failed with error, will try again in " + waitTime + " seconds.");
+                    TimeUnit.SECONDS.sleep(waitTime);
+                }
+            }
+        }
```

Comment:
I don't think that we should implement retry in the plugin.  If you cannot resolve the infrastructure problems at your site that cause the retries, then I recommend using retry in the Pipeline definition rather than placing it inside the git plugin.

If retry can't be implemented for your case in the Pipeline, then I expect one or more test cases that show how retry does not work in Pipeline implementation and how this retry implementation resolves the issue.  

Pipeline already includes a higher level retry facility that we use regularly on ci.jenkins.io for unreliable cloud infrastructure (like AWS spot instances).

Reply:
Sorry, IDE had auto-formatting on the code.